kythe/go/localrun/localrun.go

```diff
@@ -509,36 +434,65 @@ func (r *Runner) Serve(ctx context.Context) error {
 	return nil
 }
 
-// threadsafeBuffer implements a threadsafe form of bytes.Buffer
-type threadsafeBuffer struct {
-	b bytes.Buffer
-	m sync.RWMutex
+type indexer interface {
+	run(ctx context.Context, kzips []string) error
 }
 
-// Read implements Buffer.Read.
-func (b *threadsafeBuffer) Read(p []byte) (n int, err error) {
-	b.m.RLock()
-	defer b.m.RUnlock()
-	return b.b.Read(p)
+type serialIndexer struct {
+	besFile      string
+	Languages    LanguageSet
+	KytheRelease string
+	WorkingDir   string
 }
 
-// Write implements Buffer.Write.
-func (b *threadsafeBuffer) Write(p []byte) (n int, err error) {
-	b.m.Lock()
-	defer b.m.Unlock()
-	return b.b.Write(p)
-}
+func (si *serialIndexer) run(ctx context.Context, kzips []string) (io.Reader, error) {
+	indexedOut := &bytes.Buffer{}
 
-// String implements Buffer.String.
-func (b *threadsafeBuffer) String() string {
-	b.m.RLock()
-	defer b.m.RUnlock()
-	return b.b.String()
-}
+	fmt.Fprintf(os.Stderr, "Beginning serial indexing\n")
+
+	for i, k := range kzips {
+		log := func(i int, m string, v ...interface{}) {
+			den := len(kzips)
+			fmt.Fprintf(os.Stderr, "[%d/%d] %v \n", i, den, fmt.Sprintf(m, v...))
+		}
+
+		log(i, "Started indexing %q", k)
+
+		if f, err := os.Stat(k); err != nil {
+			log(i, "file error: %v", err)
+		} else if f.Size() == 0 {
+			log(i, "skipping empty .kzip")
+			continue
+		}
+
+		parts := strings.Split(k, ".")
+		langStr := parts[len(parts)-2]
+		l, ok := LanguageMap[langStr]
+		if !ok {
+			log(i, "Unrecognized language")
+			return indexedOut, fmt.Errorf("unrecognized language: %v", langStr)
+		}
+
+		cmdCtx, cancel := context.WithTimeout(ctx, 300*time.Second)
```

Comment:
Make this a constant or a flag. 10 minutes might be a safer default but really depends on if you want to give up early on big compilation units or try to let them finish.

Reply:
Done.